src/Oscoin/Crypto/Address/Internal.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE UndecidableInstances #-}
+module Oscoin.Crypto.Address.Internal where
+
+import           Oscoin.Prelude
+
+import           Oscoin.Configuration
+import           Oscoin.Crypto.Hash
+import           Oscoin.Crypto.PubKey
+
+import           Codec.CBOR.ByteArray (ByteArray)
+import           Data.Tagged
+
+{------------------------------------------------------------------------------
+  Types
+------------------------------------------------------------------------------}
+
+-- | An 'Address' is broken down as follows:
+--
+-- * /5/ bytes for the version prefix, so allocated:
+--
+-- @| 00 | 00 | 00 | 00 | 00 |
+--   ^    ^     ^   ^    ^
+--   |    |     |   |    |_ reserved (unused)
+--   |    |     |   |_ reserved (unused)
+--   |    |     |_ address format (only cbor for now)
+--   |    |_ address type (mainnet, testnet, devnet, multisig, etc)
+--   |_  protocol version (bumped on hard forks)@
+--
+-- * /23/ bytes for the CBOR-encoded address payload
+--
+-- * /6/ bytes for the checksum, computed from the @prefix + payload@.
+--
+data Address c = Address
+    { addressPrefix  :: AddressPrefix
+    , addressPayload :: AddressPayload c
+    } deriving (Eq, Show)
+
+data AddressPrefix = AddressPrefix
+    { protocolVersion :: ProtocolVersion
+    , addressType     :: AddressType
+    , addressFormat   :: AddressFormat
+    } deriving (Eq, Show)
+
+-- | A 'ProtocolVersion' is a byte representing the current version of the
+-- Oscoin protocol running. This makes possible to issue forks of the system
+-- by simply acting on this value, so that nodes running the old protocol
+-- won't be able to use addresses from the new fork. Conversely, nodes
+-- running the fork can be programmed to support both formats.
+newtype ProtocolVersion =
+    ProtocolVersion ProtocolVersionTag
+    deriving (Eq, Show)
+
+data ProtocolVersionTag =
+      ProtocolVersion_V1
+      deriving (Eq, Show)
+
+-- | The 'AddressType', i.e. the qualifying type of this 'Address'.
+-- Example:
+--
+-- 0x01 Mainnet
+-- 0x02 Testnet
+-- 0x03 Devnet
+-- ..
+newtype AddressType =
+    AddressType AddressTypeTag
+    deriving (Eq, Show)
+
+data AddressTypeTag =
+      AddressTypeMainnet -- (Mainnet, Production)
+    | AddressTypeTestNet -- (Testnet, Production)
+    | AddressTypeDevnet  -- (Devnet,  Production)
+    deriving (Eq, Show)
+
+-- | The 'AddressFormat' allows for flexibility in the 'Address' payload, i.e.
+-- it specifies the serialisation format used to encode the 'AddressPayload'.
+-- For now we have simply:
+--
+-- 0x01 -- CBOR
+--
+newtype AddressFormat =
+    AddressFormat AddressFormatTag
+    deriving (Eq, Show)
+
+data AddressFormatTag =
+    AddressFormatTagCBOR
+    deriving (Eq, Show)
+
+-- | The 'AddressPayload' is simply the hash of a 'PublicKey'.
+--
+-- NOTE(adn): The use of 'Tagged' here is due to the fact we cannot encode
```

Comment:
General comment on these notes: Since they are meant for haddocks, I don't think it makes sense to "sign" them :grin:

Reply:
ha, good point. It's actually quite a shame that haddocks renders them ( we might possibly hack around that by putting a white space between the "real" comment and the note). I guess signing is an habit which makes one accountable to either cleanup (if it's a FIXME) or it serves as a point of contact for other developers wanting to shed some lights on the comment itself (like: who the hell wrote this piece of crap? Oh, it's Alfredo 😅 ).

I do concur though for Haddock this is a no go. Do you have any tips? Embed them _inside_ the function? 🤔

Reply:
I think if it's intended to be cleaned up, then I'd keep the signature, but put it *inside* the function. If it's intended as a permanent note, I'd keep it in the haddock and remove the signature. There's some examples of this in the codebase, we usually use `Nb. Here is my awesome note` style, as it looks nicer in the haddocks.

Reply:
I would say this is an internal, permanent note, so it might make sense to put it inside. However, I do have a clear example of a "awesome note" style 😁 : in `fmtAddress`, when we warn the user about the function being heavy, that's exactly the kind of "unsigned" `Nb.` content we want to share on Haddock (methinks).

Reply:
35cf933